Linearize rm-750 audio driver volume control

mha1 opened the pull request:
Please set volume control to  Special Function Volume, Source Pot S1 or S2 and observe volume for different Pot positions. Compared to the TX16s the TX15 volume control is very non linear compared to the TX16s. The TX16s has a nice linear volume output from Min (Off), Center (half volume), Max (max Volume). The TX15 has nearly no volume increase going from Min to Center with Center having just a little volume far from half volume. Nearly all the volume range takes place from center to Max.

Summary of changes:
- introduced a linearization curve
- integrated zero volume value in curve
- integrated 90% volume limit in curve

<img width="1070" height="605" alt="image" src="https://github.com/user-attachments/assets/d6395a14-34a0-4d19-b02f-877e51383138" />

mha1 committed:
Linearize TX15 audio volume

3djc commented:
Actually I much prefer it the way it currently is, where you have fine control when at home/night, and when in open air, you just crack it up. The current scale is actually linear in dB

raphaelcoeffic reviewed (dismissed):
Since this change is highly related to a particular board using this driver, I would expect the lookup table to be located in the TX15 specific code and not in the generic TAS2505 driver, which might be used later by others as well.

mha1 committed:
moved lookup table to TX15 specific audio_driver to keep TAS2506 driver generic.

mha1 committed:
cleanup

mha1 commented:
> Don't you mean logarithmic? As volume is almost universally adjusted in a logarithmic manner.

Yes, it's more like trying to achieve a db(a) characteristic that fits the human ear's perception of volume.

Have you tried a before/after with the TX15?

3djc commented:
Previous radio (but X12S and possibly some Flysky, I do not know) use a audio routine that dates AVR time, where doing logarithm calculations was just a dream if you wanted to maintain performance, so volume is linear on those (https://github.com/EdgeTX/edgetx/blob/6c1af5a343b71a356acb9fc8cfc137ac4deef418/radio/src/audio.cpp#L658)

TX15 does add a dedicated audio amplifier with volume control, which logically applies a dB scale for volume control.

So are they different ? Yes
Was TX16S ideal ? Nop a simple basic linear volume

If anything, since we can do more calculation on F4 processors than AVR times, it is those older radio we should modify to have logarithmic volume control, not disable it on more performing radio

mha1 committed:
Merge branch 'main' into TX15-volume-adjust

mha1 committed:
Merge branch 'main' into TX15-volume-adjust

J-Sorenson commented:
Is there a way to get a community vote on this?  I prefer the current logarithmic audio scale on the new TX15 compared to my old Zorro.  I don't want to add _yet another_ feature by suggesting we make the scale selectable (the GUI is cluttered enough).  I figure those that want to assign volume to a slider and make it "just right" can use a curve.  Granted, we don't have global curves, so it would have to be copied to each model or made into a Lua script.

The problem is that if this were to get merged, would I lose the resolution of setting the lower volume?  I can't fix that with a Lua script.

mha1 commented:
> The problem is that if this were to get merged, would I lose the resolution of setting the lower volume? I can't fix that with a Lua script.

No you wouldn't. Look at both tables (old vs new) and/or the graph. Both use the same range bit with different distribution.

J-Sorenson commented:
> > The problem is that if this were to get merged, would I lose the resolution of setting the lower volume? I can't fix that with a Lua script.
> 
> No you wouldn't. Look at both tables (old vs new) and/or the graph. Both use the same range bit with different distribution.

I said resolution (step size), not limits. Perhaps this issue could hang for a while until more new owners chime in.

J-Sorenson commented:
> 
> Exactly the same resolution. 24 volume steps (including zero volume)

Of course it's always 24 "steps". The point of this issue is what decibels each step should be at the upper vs lower end. Look, your concern is that something changed from what you are familiar with. My point is that the new divisions are better.  We don't have a clear consensus on what the final curve should be unless we can get more feedback. It won't be decided between us two.

mha1 commented:
Btw, upper and lower level are the same too. Max volume is the same max volume as the current implementation, off is off in both too.

mha1 committed:
feat(rm-h750): linearize TX15 audio volume using lookup table

Move volume linearization lookup table from the generic TAS2505/2506
driver into the TX15-specific audio_driver to keep the driver generic.

mha1 committed:
Merge branch 'TX15-volume-adjust' of github.com:mha1/edgetx into TX15-volume-adjust

mha1 committed:
Merge branch 'main' into TX15-volume-adjust

mha1 committed:
Changes apply to all radios using this audio driver.

mha1 committed:
Merge branch 'main' into TX15-volume-adjust

OzDruiD commented:
Pretty impressed with this  - top and bottom may technically be the same. but 'feels' better\more consistent.
Volume at mid-detent is amazing - and whilst sounds slighly different to Mk2 - as expected with diff electronics\speaker\radio accoustics,  the implementation is FAR better than current method.

Please please normalise this into the build - or if necessary make it an audio option we can have ON \ OFF and not a rebuild flag each release.   Dropped some before\after videos into EdgeTX discord as well.  https://discord.com/channels/839849772864503828/839849772864503832/1498268287123984485

VancouverUmbrella commented:
Hey folks,

I was directed to this ticket when I complained on RCG about the sound quality of my new RM TX16S Mk3. 

Like others who've posted here, I put the sound volume control on a pot via a Special Function. With that setup, I was surprised to discover that I needed to crank the pot to about 75% in order to get an equivalent volume to my FrSky QX7 with a pot set to about 40%. I'm definitely in favour of changing the behaviour of the volume control in the Mk3.

Can I ask something further? Specifically, is there any possible connection between the issue in this ticket and what I find to be disappointing audio quality from my Mk3? The sound from the Mk3 seems muffled and hissy compared to my QX7. It's not a subtle difference. It surprised me from the moment I first booted up my Mk3. Perhaps it's down to the properties of audio amp or speaker hardware?

A huge thank you to all who put such care and dedication into EdgeTX.

3djc commented:
> Can I ask something further? Specifically, is there any possible connection between the issue in this ticket and what I find to be disappointing audio quality from my Mk3?

Absolutely not. This does not affect the quality of the sound produced in any way shape or form. It just affects the volume curve, it does not affect maximum volume either. On those new radio it is made to follow better human hear characteristics (https://support.rationalacoustics.com/support/solutions/articles/150000214511-linear-and-logarithmic-scaling), like any other modern audio device on the market

Mike-HRC commented:
I also have this weirdness on my TX15 - almost silent until half - way, and then audible at about 65%.  Please change in some way so that it is more "normal"

VancouverUmbrella commented:
> > Can I ask something further? Specifically, is there any possible connection between the issue in this ticket and what I find to be disappointing audio quality from my Mk3?
> 
> Absolutely not. This does not affect the quality of the sound produced in any way shape or form. It just affects the volume curve, it does not affect maximum volume either. 

Thank you. Very helpful.
